internal/errors/error.go

```diff
@@ -62,19 +65,68 @@ func E(opt ...Option) error {
 		code = opts.withCode
 	}
 
-	return &Err{
+	err = &Err{
 		Code:    code,
 		Op:      opts.withOp,
 		Wrapped: opts.withErrWrapped,
 		Msg:     opts.withErrMsg,
 	}
+	if opts.withoutEvent {
+		return err
+	}
+
+	event.WriteError(ctx, "", err)
+
+	return err
 }
 
 // New creates a new Err with provided code, op and msg
 // It supports the options of:
 //
 // * WithWrap() - allows you to specify an error to wrap
-func New(c Code, op Op, msg string, opt ...Option) error {
+func New(ctx context.Context, c Code, op Op, msg string, opt ...Option) error {
+	if c != Unknown {
+		opt = append(opt, WithCode(c))
+	}
+	if op != "" {
+		opt = append(opt, WithOp(op))
+	}
+	if msg != "" {
+		opt = append(opt, WithMsg(msg))
+	}
+
+	return E(ctx, opt...)
+}
+
+// EDeprecated is the legacy version of E which does not
+// create an event. Please refrain from using this.
+func EDeprecated(opt ...Option) error {
```

Comment:
Others may feel differently but instead of doing a phased deprecation of these internal functions, why not just make it a refactor and add the context at the caller site?  Where there is no context available we could use context.TODO().  That function's documentation says "Code should use context.TODO when
 it's unclear which Context to use or it is not yet available (because the surrounding function has not yet been extended to accept a Context parameter)."  which seems to fit our case.

Reply:
Hum... so there would seem to be a significant side effects from your suggested approach: none of the error events with context.TODO would have an associated request id which really is a broken state for observability and not something we'd likely want for 50% of our error events.  Also, this is an exported function and it's used in quite a few places outside the errors pkg.

Reply:
E and New should handle cases where ctx are passed to it w/o a request id present.  My suggested approach  doesn't seem any more broken than not calling the observability enabled error functions at all which is the approach this change implements.  This approach seems to ensure that EDeprecated and NewDeprecated will continue to exist forever given that it is used in so many places and there are places where requestId will never be present.

Reply:
I tend to agree with Todd; once we have fully migrated all errors, what do we want to do with errors generated outside the scope of a `RequestId`?  It seems that the event package foresaw the possibility of not having an eventer in the context and therefore uses the `SysEventer`.  It seems sane to use `context.TODO`

Another possibility is add the context to `New` and not `E`.  `E` was always meant to give the dev more granularity in error creation (i.e. errors that do not have a Message).  The usage of `E` is extremely sparse so we could have it be purely used for error creation.  

Something else to consider is there are a couple of usages of `New` and possibly `E` where we also call the `WithWrap` option, we could therefore have multiple `WriteError` events being called along the error path.

Reply:
We have accommodated errors that aren't associated with a request within WriteError(...) so there is no requirement that `NewDeprecated` and `EDeprecated` exist forever.    BTW, I wasn't try to suggest that every error must have a request id.  I believe what we're trying to do is making it explicitly clear in the code base everywhere we created errors without regard to their context.  In a few months, when someone looks at `errors.New` or `errors.E` with a context.TODO  they won't have to reason if it was intentional or just something missed when we tried to clean up all the bits during the refactor.

Reply:
Could we add the deprecation steps in the documentation for the new deprecated functions so we know 
1. what needs to be done to remove them and
2. we know when they are safe to remove?

Bonus points for creating jira tasks for the deprecation steps :)

Reply:
Happy to add some comments and Jira tasks about the steps needed to remove the deprecated functions.